File: Leetcode/82-Remove-Duplicates-from-Sorted-List-II/RemoveDuplicatesFromSortedListII.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    if (!head || !head.next) {
        return head
    }
    var dummy = new ListNode(0)
    dummy.next = head
    var cur = dummy
    var isDup = false
    while (cur && cur.next) {
        if (cur.next.next && cur.next.val == cur.next.next.val) {
            isDup = true
            cur.next.next = cur.next.next.next
        } else {
            if (isDup) {
                isDup = false
                cur.next = cur.next.next
            } else {
                cur = cur.next
            }
        }
    }
    return dummy.next
};

File: Leetcode/82-Remove-Duplicates-from-Sorted-List-II/RemoveDuplicatesFromSortedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* cur = dummy;
        bool isDuplicate = false;
        while (cur!=NULL && cur->next!=NULL) {
            if (cur->next->next!=NULL && cur->next->val == cur->next->next->val) {
                isDuplicate = true;
                cur->next->next = cur->next->next->next;
            } else {
                if (isDuplicate) {
                    isDuplicate = false;
                    cur->next = cur->next->next;
                } else {
                    cur = cur->next;
                }
            }
        }
        return dummy->next;
    }
};
